File: GFG POTD/022-September-2025/020-longest-subarray-length.java
// https://www.geeksforgeeks.org/problems/longest-subarray-length--202010/1/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(1)
    public static int longestSubarray(int[] arr) {
        int n = arr.length;
        
        int[] nge = getNge(arr);
        int[] pge = getPge(arr);
        
        int maxLen = 0;
        for (int i = 0; i < n; i ++) {
            int len = (nge[i] - pge[i] - 1);
            if (arr[i] <= len) {
                maxLen = Math.max(maxLen, len);
            }
        }
        return maxLen;
    }
    
    // TC : O(N)
    // SC : O(1)
    private static int[] getNge(int[] arr) {
        int n = arr.length;
        Stack<Integer> stk = new Stack<>();
        int[] result = new int[n];
        Arrays.fill(result, n);
        for (int i = n - 1; i >= 0; i --) {
            while (!stk.isEmpty() && arr[stk.peek()] <= arr[i]) {
                stk.pop();
            }
            if (!stk.isEmpty()) result[i] = stk.peek();
            stk.push(i);
        }
        return result;
    }

    // TC : O(N)
    // SC : O(1)
    private static int[] getPge(int[] arr) {
        int n = arr.length;
        Stack<Integer> stk = new Stack<>();
        int[] result = new int[n];
        Arrays.fill(result, -1);
        for (int i = 0; i < n; i ++) {
            while (!stk.isEmpty() && arr[stk.peek()] <= arr[i]) {
                stk.pop();
            }
            if (!stk.isEmpty()) result[i] = stk.peek();
            stk.push(i);
        }
        return result;
    }
}

File: GFG POTD/022-September-2025/020-longest-subarray-length.cpp
// https://www.geeksforgeeks.org/problems/longest-subarray-length--202010/1/




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    // TC : O(N)
    // SC : O(N)
    vector<int> getNge(vector<int> &arr) {
        int n = arr.size();
        stack<int> stk;
        vector<int> result(n, n);
        for (int i = n - 1; i >= 0; i --) {
            while (!stk.empty() && arr[stk.top()] <= arr[i]) {
                stk.pop();
            }
            if (!stk.empty()) result[i] = stk.top();
            stk.push(i);
        }
        return result;
    }
    
    // TC : O(N)
    // SC : O(N)
    vector<int> getPge(vector<int> &arr) {
        int n = arr.size();
        stack<int> stk;
        vector<int> result(n, -1);
        for (int i = 0; i < n; i ++) {
            while (!stk.empty() && arr[stk.top()] <= arr[i]) {
                stk.pop();
            }
            if (!stk.empty()) result[i] = stk.top();
            stk.push(i);
        }
        return result;
    }
  public:
    // TC : O(N)
    // SC : O(N)
    int longestSubarray(vector<int>& arr) {
        int n = arr.size();
        
        vector<int> nge = getNge(arr);
        vector<int> pge = getPge(arr);
        
        int maxLen = 0;
        for (int i = 0; i < n; i ++) {
            int len = (nge[i] - pge[i] - 1);
            if (arr[i] <= len) {
                maxLen = max(maxLen, len);
            }
        }
        return maxLen;
    }
};